src/python/gudhi/datasets/generators/_points.cc

```diff
@@ -5,117 +5,115 @@
  *    Copyright (C) 2021 Inria
  *
  *    Modification(s):
+ *      - 2025/01 Vincent Rouvreau: Use nanobind instead of PyBind11 for python bindings
  *      - YYYY/MM Author: Description of the modification
  */
 
-#include <pybind11/pybind11.h>
-#include <pybind11/numpy.h>
+#include <vector>
 
-#include <gudhi/random_point_generators.h>
-#include <gudhi/Debug_utils.h>
+#include <nanobind/nanobind.h>
+#include <nanobind/ndarray.h>
+#include <nanobind/stl/string.h>
 
 #include <CGAL/Epick_d.h>
 
-namespace py = pybind11;
-
-
-typedef CGAL::Epick_d< CGAL::Dynamic_dimension_tag > Kern;
-
-py::array_t<double> generate_points_on_sphere(size_t n_samples, int ambient_dim, double radius, std::string sample) {
-
-    if (sample != "random") {
-        throw pybind11::value_error("This sample type is not supported");
-    }
-
-    py::array_t<double> points({n_samples, (size_t)ambient_dim});
-
-    py::buffer_info buf = points.request();
-    double *ptr = static_cast<double *>(buf.ptr);
+#include <gudhi/random_point_generators.h>
+#include <gudhi/Debug_utils.h>
 
-    GUDHI_CHECK(n_samples == buf.shape[0], "Py array first dimension not matching n_samples on sphere");
-    GUDHI_CHECK(ambient_dim == buf.shape[1], "Py array second dimension not matching the ambient space dimension");
+namespace nb = nanobind;
 
+typedef CGAL::Epick_d<CGAL::Dynamic_dimension_tag> Kern;
 
-    std::vector<typename Kern::Point_d> points_generated;
+auto generate_points_on_sphere(const size_t n_samples, const int ambient_dim, double radius, std::string sample)
+{
+  if (sample != "random") {
+    throw nb::value_error("This sample type is not supported");
+  }
 
-    {
-        py::gil_scoped_release release;
-        points_generated = Gudhi::generate_points_on_sphere_d<Kern>(n_samples, ambient_dim, radius);
-    }
+  std::vector<typename Kern::Point_d> points_generated;
+  {
+    nb::gil_scoped_release release;
+    points_generated = Gudhi::generate_points_on_sphere_d<Kern>(n_samples, ambient_dim, radius);
+  }
 
-    for (size_t i = 0; i < n_samples; i++)
-        for (int j = 0; j < ambient_dim; j++)
-            ptr[i*ambient_dim+j] = points_generated[i][j];
+  // Reserve sufficient memory space to copy data
+  std::vector<double> points(n_samples * ambient_dim);
+  double *ptr = points.data();
+  for (size_t i = 0; i < n_samples; i++)
+    for (int j = 0; j < ambient_dim; j++) ptr[i * ambient_dim + j] = points_generated[i][j];
 
-    return points;
+  return nb::ndarray<nb::numpy, double>(points.data(), {n_samples, (size_t)ambient_dim}).cast();
```

Comment:
IIUC we do 2 copies, from a vector of CGAL points to a vector of double to a ndarray. It feels like we could skip the intermediate vector.
(the performance of this function is not important, I am more trying to understand how nanobind works)

Reply:
Yes, we could remove the `.cast()` at the end of ndarray definition. Then no copy is made other than the copy of the pointer `points.data()`. But then, we need to precise ndarray that it takes over ownership, by making a `new` instead of a local declaration (to avoid `points` to be destroyed at the end of the method) and telling it to destroy the pointer when the array is not used anymore.